x-pack/plugins/observability_solution/apm/public/components/app/service_logs/index.tsx

```diff
@@ -84,3 +149,99 @@ export function getInfrastructureKQLFilter({
 
   return [serviceNameAndEnvironmentCorrelation, ...containerIdCorrelation].join(' or ');
 }
+
+export function getInfrastructureFilter({
+  containerIds,
+  environment,
+  serviceName,
+}: {
+  containerIds: string[];
+  environment: string;
+  serviceName: string;
+}): QueryDslQueryContainer {
+  return {
+    bool: {
+      should: [
+        ...getServiceShouldClauses({ environment, serviceName }),
+        ...getContainerShouldClauses({ containerIds }),
+      ],
+      minimum_should_match: 1,
+    },
+  };
+}
+
+export function getServiceShouldClauses({
+  environment,
+  serviceName,
+}: {
+  environment: string;
+  serviceName: string;
+}): QueryDslQueryContainer[] {
+  const serviceNameFilter: QueryDslQueryContainer = {
+    term: {
+      [SERVICE_NAME]: serviceName,
+    },
+  };
+
+  if (environment === ENVIRONMENT_ALL.value) {
+    return [serviceNameFilter];
+  } else {
+    return [
+      {
+        bool: {
+          filter: [
+            serviceNameFilter,
+            {
+              term: {
+                [SERVICE_ENVIRONMENT]: environment,
+              },
+            },
```

Comment:
We could use helper functions to build the filer clause. Wdyt?

```ts
...termQuery(SERVICE_ENVIRONMENT, environment),
```

Same for the `exists` below

```
existsQuery(SERVICE_ENVIRONMENT)